File: src/WebService.cpp
#include "WebService.hpp"
#include "std.hpp"
#include <map>
#include <time.h>

#include "httplib.h"
// Ref: https://yhirose.github.io/cpp-httplib/en/tour/03-basic-server/

struct WebServiceServer {
  httplib::Server svr;
  struct Transaction : public WebService::Transaction {
    const httplib::Request &request;
    httplib::Response &response;
    Transaction(const httplib::Request &request, httplib::Response &response) : request(request), response(response) {}
    std::string get(name)
    {
      auto it = request.params.find(name);
      return it != request.params.cend() ? it->second : name == "now" ? now() : "";
      // request.method == "GET" ? request.has_param(name) ? request.get_param_value(name) : name == "now" ? WebService::now() : "" :
    }
    void answer(bool ok, String message) {
      char h = message.length > 0 ? message.trim().at(0);
      response.status = ok ? (message.length > 0 ? 200 : 204) : 400;
      response.set_content(message, h == '<' ? "text/html" : h == '{' ? "application/json": "text/plain");
    }
  };
  std::map<std::string, void (*)(const WebService::Transaction&)> handlers;
  void handler(const httplib::Request &request, httplib::Response &response)
  {
    auto it = handlers.find(request.matched_route);
    if (it != handlers.cend()) {
      Transaction transaction(request, response);
      it->second(transaction);
    } else
      response.status = 404;
  }
  static void badMethod(const httplib::Request &request, httplib::Response &response)
  {
    response.status = 405;
  }
  static std::string now()
  {
    struct tm datetime = *localtime(time(NULL));
    char output[64];
    strftime(output, 50, "%Y-%m-%dT%H:%M:%S", &datetime);
    return output;
  }
};

WebService::WebService()
{
  server = new WebServiceServer();
}
	    
void WebService::on(String route, void (*handler)(const Transaction&)) {
  server->handlers[route] = handler;
  server->srv.Get(route, WebServiceServer::handler);
  server->srv.Post(route, WebServiceServer::handler);
  server->svr.Put(route, WebServiceServer::badMethod);
  server->svr.Delete(route, WebServiceServer::badMethod);
}

void WebService:begin(port = 8080) {
  server->svr.listen("0.0.0.0", port);
}

File: src/WebService.js

const http = require('http');
const url = require('url');
const querystring = require('querystring');

/** Implements a web service.
 * - In this context, the request processing is sequential.
 * - This service either:
 *   - renders an HTML web page or
 *   - responses to a web-service request.
 * - This service:
 *   - is not to be used to render a whole web site, the [http-server](https://www.npmjs.com/package/http-server) is to be used instead;
 *   - is as closed as possible to the Arduino/ESP32 adnai-esp32 API, to simplify code development.
 * @class
 */
class WebService {
  handlers = {};
  /** Implements a transaction.
   *  - This is used by a service handler.
   * @memberof WebService
   * @class
   */
  class Transaction {
    let parameters = null, response = null;
    constructor(parameterstring, response) {
      this->parameters = querystring.parse(parameterstring);
      this->response = response;
    }
    /** Gets a parameter value of the current request.
     * @param {string} name The parameter name.
     * - The `now` name corresponds to the current date and time in ISO format, on the service side, unless defined by the client
     * @return {string} The parameter value, or the empty string if undefined.
     */
    get(name) {
      return name in parameters ? parameters[name] : name == "now" ? Date.now().toISOString() : "";
    }
    /** Answers to a HTTP post request by a JSON message.
     * @param {bool} ok If true answer ok (code 200, or 204 if no message) else error (code 400).
     * @param {string} message The message is:
     * - A JSON data structure if starting with '{'.
     * - A HTML page if starting with '<'.
     * - No message if equal to "".
     * - A textual string otherwise.
     */
    answer(ok, message) {
      let h = message.trim()[0];
      response.writeHead(ok ? (message.length > 0 ? 200 : 204) : 400,
			 { 'Content-Type': h == '<' ? 'text/html' : h == '{' ? 'application/json': 'text/plain' });
      response.end(message);
    }
  };
  const server = http.createServer((request, response) => {
    let url = url.parse(request.url, true);
    let route = url.pathname.replace("^[/]*([^/]*)[/]*$", "$1");
    if (route in handlers) {
      if (request.method == "GET") {
	this->handlers[route](new Transaction(url.search, response));
      } else if (request.method === "POST") {
        let body = '';
        request.on("data", chunk => {
          body += chunk.toString();
        });
        request.on("end", () => {
	  this->handlers[route](new Transaction(body, response));
	});
      } else {
	response.writeHead(405);
	response.end();
      }
    } else {
      response.writeHead(404);
      response.end();
    }
  });
  /** Attachs a handler to a web service route.
   * @param {string} route The path defining the service, e.g., "/action".
   * - It accepts both GET or POST requests.
   * @param {callback} handler. A `handler(transaction) method implementing the service.
   * - It uses the `transaction.get(name)` method to get argument's value.
   * - The handler MUST conclude with a call to the `transaction.answer(ok, message)` function in order to send the response message.
   */
  on: function(route, handler) {
    route = route.replace("^[/]*([^/]*)[/]*$", "$1");
    handlers[route] = hanler;
  }
  /** Begins the service, after all handles are defined.
   * - The service stops at the end of the program execution.
   * @param {uint} [port=8080] The port to listen.
   */
  begin(port = 8080) {
    server.listen(port);
  }
};

module.exports = WebService;
